### PR TITLE
Support python wheel installation

### PR DESCRIPTION
**Pull request by swimdi**
Hi all:
this PR is to support wheel installation and try to minimize code change. The related issue is #6576

The wheel no need to export `ARCH_NAME` & `TT_METAL_HOME` & `PYTHONPATH` & `TT_METAL_ENV` and can run the example on readme success
```
import ttnn
import torch

with ttnn.manage_device(device_id=0) as device:
   a = torch.ones((5, 7))
   b = torch.ones((1, 7))

   a = ttnn.from_torch(a, device=device, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT)
   b = ttnn.from_torch(b, device=device, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT)

   output = a + b
   output = ttnn.to_torch(output)
```
After installation, the folder structure on the site-package is
```
lib/python3.8/site-package
├── tt_lib
│   ├── build/
│   │   ├── hw/toolchain/*.ld => for createKernel use
│   │   └── build/lib/*.so => pybind so for import
│   ├── models/
│   ├── tt_metal@ -> ../tt_metal
│   └── tt_eager/tt_dnn
├── tt_metal
└── ttnn
```
Because want the code change minimze, so it treat `lib/python3.8/site-package/tt_lib` as `TT_METAL_HOME`,so the folder structure of `lib/python3.8/site-package/tt_lib` make like the struture of `tt-metal` repo. If `tt_lib/tt_metal` softlink not exist, then will occur this error
```
RuntimeError: Error: device descriptor file xxx/venv/lib/python3.8/site-packages/tt_lib/tt_metal/soc_descriptors/grayskull_120_arch.yaml does not exist!
```
c.c. @tt-rkim @TT-billteng @dimitri-tenstorrent

### CODE REVIEW

**Comment by swimdi** [comment]
I've solve the merge conflict and it also can pass the example on readme success
```
import ttnn
import torch

with ttnn.manage_device(device_id=0) as device:
   a = torch.ones((5, 7))
   b = torch.ones((1, 7))

   a = ttnn.from_torch(a, device=device, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT)
   b = ttnn.from_torch(b, device=device, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT)

   output = a + b
   output = ttnn.to_torch(output)

print(output)
```

File: pyproject.toml [hunk 1]
```diff
@@ -22,6 +22,7 @@ dependencies = [
   "Pillow==10.3.0",
   "jupyterlab==4.0.7",
   "ipywidgets==8.1.1",
+  "pytest==7.2.2",
```
Comment (dimitri-tenstorrent) [inline pyproject.toml:25]: Why do we need pytest inside of `pyproject.toml`? We have another file that handles dev dependencies in `tt_metal/python_env/requirements-dev.txt`

Comment (swimdi) [reply pyproject.toml]: I missed to delete this code. When I tried the main branch on March 26 there needs the pytest, but after I rebased to the current main branch and test, it don't needs this code.
I have deleted it, thanks for you reviewing.

File: ttnn/module.mk
Comment (tt-rkim) [inline ttnn/module.mk:1]: Unnecessary changes for the makefile, please delete as we now use a related option in CMake and properly handle creating a static lib

File: ttnn/ttnn/__init__.py [hunk 2]
```diff
@@ -15,6 +15,10 @@
 import tt_lib as _tt_lib
 import ttnn._ttnn
 
+import site
+
+site.addsitedir(pathlib.Path(__file__).parent.parent / "tt_lib")
```
Comment (tt-rkim) [inline ttnn/ttnn/__init__.py:20]: Is this necessary if we have `tt_lib` available as an importable package now?

File: ttnn/ttnn/__init__.py [hunk 3]
```diff
@@ -127,11 +131,43 @@ def save_config_to_json_file(json_path):
         save_config_to_json_file(CONFIG_PATH)
 
 
+def _check__ttnn_so_rpath():
```
Comment (tt-rkim) [inline ttnn/ttnn/__init__.py:134]: Please make this an importable function from `tt_lib` instead, with parameters for:

- `so_name` or something similar, for usage in `f.startswith(so_name)`
- `new_path` or something similar

File: setup.py [hunk 4]
```diff
@@ -107,6 +107,19 @@ def get_build_env():
             **os.environ.copy(),
             "TT_METAL_HOME": Path(__file__).parent,
             "TT_METAL_ENV": "production",
+            # Need to create static lib for tt_metal runtime because currently
```
Comment (tt-rkim) [inline setup.py:110]: This comment's information is correct, however I would you like to reword it. Please change it to the following:

```
"""
Currently, the ttnn (ttnn/_ttnn.so) and tt_lib (tt_lib/_C.so)
both link to the tt_metal runtime. The specific thing in
ttnn linking to tt_metal is likely the implementation of
ttnn.manage_device.

However, because of the singleton design of
tt_cluster in tt_metal, both tt_lib and ttnn will have a
copy of the cluster object, causing a hang during
device operations in ttnn, such as calling
output = ttnn.to_torch(output).

Ultimately, we will not statically build tt_metal, and
opt to dynamically set rpath of the bindings to the
packaged libs for now.
"""
```

File: INSTALLING.md [hunk 5]
```diff
@@ -119,7 +119,7 @@ Please follow the next additional steps if you want to contribute to the codebas
 1. Install dependencies
 
 ```sh
-sudo apt install clang-6.0=1:6.0.1-14 git git-lfs cmake=3.16.3-1ubuntu1.20.04.1 pandoc libtbb-dev libcapstone-dev pkg-config
+sudo apt install clang-6.0=1:6.0.1-14 git git-lfs cmake=3.16.3-1ubuntu1.20.04.1 pandoc libtbb-dev libcapstone-dev pkg-config patchelf
```
[the remark this answers is not in the document]
Comment (swimdi) [reply INSTALLING.md]: ok, I had added it

File: tt_eager/tt_lib/__init__.py [hunk 6]
```diff
@@ -1,5 +1,80 @@
...
+    if os.path.exists(check_f):
+        return
+    target_so = None
+    for f in os.listdir(directory):
+        if f.startswith("_C") and f.endswith(".so"):
+            target_so = directory / f
+            break
+    if not target_so:
+        return
+    import subprocess
```
[the remark this answers is not in the document]
Comment (swimdi) [reply tt_eager/tt_lib/__init__.py]: ok, I had move it

File: tt_eager/tt_lib/__init__.py [hunk 7]
```diff
@@ -1,5 +1,80 @@
...
+import os
+from pathlib import Path
 
+
+def _check__C_so_rpath():
+    directory = Path(__file__).parent
+    check_f = directory / ".rpath_checked"
+    if os.path.exists(check_f):
+        return
+    target_so = None
```
[the remark this answers is not in the document]
Comment (swimdi) [reply tt_eager/tt_lib/__init__.py:24]: Ok, I had added the logger code on it

File: tt_eager/tt_lib/__init__.py [hunk 8]
```diff
@@ -1,5 +1,80 @@
...
+
+    new_rpath = directory
+    if has_not_found(eager_so):
+        subprocess.check_call(f"patchelf --set-rpath {new_rpath} {eager_so}", shell=True)
+    if has_not_found(metal_so):
+        subprocess.check_call(f"patchelf --set-rpath {new_rpath} {metal_so}", shell=True)
+    subprocess.check_call(f"touch {check_f}", shell=True)
+
+
+sit_pkgs_tt_lib = Path(__file__).parent
```
[the remark this answers is not in the document]
Comment (swimdi) [reply tt_eager/tt_lib/__init__.py]: It's my typo, thanks for reminding

File: tt_eager/tt_lib/__init__.py [hunk 9]
```diff
@@ -1,5 +1,80 @@
...
+
+if "ARCH_NAME" not in os.environ or os.environ["ARCH_NAME"] == "":
+    os.environ["ARCH_NAME"] = "grayskull"
+if "TT_METAL_HOME" not in os.environ or os.environ["TT_METAL_HOME"] == "":
+    # Workaround: treat $SITE_PACKAGES/tt_lib as TT_METAL_HOME
```
[the remark this answers is not in the document]
Comment (swimdi) [reply tt_eager/tt_lib/__init__.py]: I removed the soft link code and set `os.environ["TT_METAL_HOME"] = str(site_pkgs_tt_lib.parent)`, but the jit build seems need the linker script from `TT_METAL_HOME/build/hw/toolchain`, so I add the soft link of `tt_lib/build` to let it pass...
```
cannot open linker script file xxx/lib/python3.8/site-packages/build/hw/toolchain/trisc2.ld: No such file or directory
```

File: tt_eager/tt_lib/__init__.py [hunk 10]
```diff
@@ -1,5 +1,80 @@
...
+    if has_not_found(target_so):
+        new_rpath = directory / "build" / "lib"
+        subprocess.check_call(f"patchelf --set-rpath {new_rpath} {target_so}", shell=True)
+    subprocess.check_call(f"touch {check_f}", shell=True)
+
+
+def _check_so_rpath_in_build_lib():
+    directory = Path(__file__).parent / "build/lib"
+    check_f = directory / ".rpath_checked"
+    if not os.path.exists(directory):
```
[the remark this answers is not in the document]
Comment (swimdi) [reply tt_eager/tt_lib/__init__.py:41]: Ok, I had added the logger code on it

File: ttnn/module.mk
Comment (swimdi) [reply ttnn/module.mk:1]: ok, I had revert the code

File: ttnn/ttnn/__init__.py (hunk 2 above)
Comment (swimdi) [reply ttnn/ttnn/__init__.py]: no, currently don't needs these code. I had deleted it

File: ttnn/ttnn/__init__.py (hunk 3 above)
Comment (swimdi) [reply ttnn/ttnn/__init__.py]: ok, I had edit it to make importable

File: tt_eager/tt_lib/__init__.py [hunk 11]
```diff
@@ -1,5 +1,79 @@
...
 # SPDX-License-Identifier: Apache-2.0
+import os
+import subprocess
+from pathlib import Path
+from loguru import logger
 
+
+def _has_not_found(target_so):
+    if not os.path.exists(target_so):
+        logger.trace(f"Shared library {target_so} not exists")
```
Comment (tt-rkim) [inline tt_eager/tt_lib/__init__.py:12]: Please change `not exists` to `does not exist`

File: tt_eager/tt_lib/__init__.py [hunk 12]
```diff
@@ -1,5 +1,79 @@
...
+    check_f = directory / f".rpath_checked_{so_name}"
+    if os.path.exists(check_f):
+        return
+    target_so = None
+    for f in os.listdir(directory):
+        if f.startswith(so_name) and f.endswith(".so"):
+            target_so = directory / f
+            break
+    if not target_so:
+        logger.trace(f"Cannot found shared library which name starts with {so_name}")
```
Comment (tt-rkim) [inline tt_eager/tt_lib/__init__.py:30]: Change `found` to `find`

File: tt_eager/tt_lib/__init__.py [hunk 13]
```diff
@@ -1,5 +1,79 @@
...
+    if _has_not_found(target_so):
+        subprocess.check_call(f"patchelf --set-rpath {new_rpath} {target_so}", shell=True)
+    subprocess.check_call(f"touch {check_f}", shell=True)
+
+
+def _check_so_rpath_in_build_lib():
+    directory = Path(__file__).parent / "build/lib"
+    check_f = directory / ".rpath_checked"
+    if not os.path.exists(directory):
+        logger.trace(f"Direcory {directory} not exists")
```
Comment (tt-rkim) [inline tt_eager/tt_lib/__init__.py:42]: `Directory {directory} does not exist`

File: tt_eager/tt_lib/__init__.py (hunk 11 above)
Comment (swimdi) [reply tt_eager/tt_lib/__init__.py]: ok, I have fixed it

File: tt_eager/tt_lib/__init__.py (hunk 12 above)
Comment (swimdi) [reply tt_eager/tt_lib/__init__.py]: ok, I have fixed it

File: tt_eager/tt_lib/__init__.py (hunk 13 above)
Comment (swimdi) [reply tt_eager/tt_lib/__init__.py]: ok, I have fixed it

**Comment by tt-rkim** [comment]
#8873 
Merged and closed.
